Add order lock for legacy redirect payments

Mayisha opened the pull request:
Fixes #1302 

In the legacy redirect payment flow, no order locking mechanism has been implemented. This might cause double order note/email issue. This PR prevents the possible double order note/email issue by imposing order locking during redirect payments to avoid processing the same order by webhook as well.

## Changes proposed in this Pull Request:
- Added order lock to the redirect payment flow for legacy checkout.

## Testing instructions
- Enable legacy checkout experience.
- Enable a few other payment methods apart from the card (i,e iDeal, EPS, Bancontact) that need redirection.
- As a shopper, make purchases with the other payment methods. Confirm that the payments work fine.

Mayisha committed:
add order lock during legacy redirect payments

Mayisha committed:
Merge branch 'develop' into task/1302-block-order-lpm

mattallan commented on includes/class-wc-stripe-order-handler.php:
```diff
@@ -82,6 +82,9 @@ public function process_redirect_payment( $order_id, $retry = true, $previous_er
 
 			WC_Stripe_Logger::log( "Info: (Redirect) Begin processing payment for order $order_id for the amount of {$order->get_total()}" );
 
+			// Lock the order.
+			$this->lock_order_payment( $order );
+
```
I think we should surround this in the same condition as the webhook handler in case the webhook comes in first.

```
if ( $this->lock_order_payment( $order ) ) {
	return;
}
```

FWIW, when testing locally my webhook was arriving before this function.

mattallan reviewed (changes requested):
Thanks @Mayisha for these changes.

I couldn't reproduce duplicate emails or order notes issues, but I did reproduce a different issue 😅 

If `process_redirect_payment()` handled the charge request but hasn't yet updated the order status, `process_webhook_payment()` will call `WC_Stripe_API::request( $this->generate_payment_request() )` which will fail with an idempotency key already used issue and throw an exception.

The exception is caught and we send an order failed email to the customer and add an order note.
But the order is processing, not failed 🙈 

![image](https://github.com/user-attachments/assets/c4863609-b817-4d1c-b821-269f0eb996a3)

It's a bit finicky but I could replicate it consistently by adding a `sleep(3)` to simulate a slow API request.

Your changes in this PR should resolve this.

---

**:x: Issue using a lock when the function retries.**

Something I just noticed while testing this is that the lock isn't cleared on retry, so if the `if ( ! empty( $response->error ) ) {` condition passes and we retry the function again, it returns early as the lock is still enabled 🤔. Leaving the order as pending.

Should we bypass the lock if `$retry` is set to true?

Mayisha committed:
Merge branch 'develop' into task/1302-block-order-lpm

Mayisha committed:
Merge branch 'develop' into task/1302-block-order-lpm

Mayisha committed:
address feedback

Mayisha committed:
unlock before retry

Mayisha commented:
Thanks @mattallan 👍 

I have addressed the feedback in [5446ef5](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3442/commits/5446ef56266eac78e59c11dd20f4c448a6c7efa9) and [4504087](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3442/commits/4504087f86f93af0855398bee52ec7ba1e17a4f9)

- Checking if the order is already locked and returning in that case.
- Unlocking the order before retrying and in the `catch` block that exits before reaching the normal unblock flow.

mattallan reviewed (approved):
Thanks @Mayisha for the latest changes!
The new commits look good. I've ran through a test again and confirmed no issues.

Happy to approve this :)

Mayisha committed:
Merge branch 'develop' into task/1302-block-order-lpm

Mayisha committed:
add changelog entry

Mayisha committed:
Merge branch 'develop' into task/1302-block-order-lpm
